Content.Client/GameObjects/Components/Instruments/InstrumentComponent.cs

```diff
@@ -0,0 +1,170 @@
+        /// </summary>
+        public event Action OnMidiPlaybackEnded;
+
+#pragma warning disable 649
+        [Dependency] private IMidiManager _midiManager;
+        [Dependency] private IFileDialogManager _fileDialogManager;
+#pragma warning restore 649
+
+        private IMidiRenderer _renderer;
+        private int _instrumentProgram = 1;
+
+        /// <summary>
+        ///     A queue of MidiEvents to be sent to the server.
+        /// </summary>
+        private Queue<MidiEvent> _eventQueue = new Queue<MidiEvent>();
+
+        /// <summary>
+        ///     Whether a midi song will loop or not.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public bool LoopMidi
+        {
+            get => _renderer.LoopMidi;
+            set => _renderer.LoopMidi = value;
+        }
+
+        /// <summary>
+        ///     Changes the instrument the midi renderer will play.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public int InstrumentProgram
+        {
+            get => _instrumentProgram;
+            set {
+                _instrumentProgram = value;
+                _renderer.MidiProgram = _instrumentProgram;
+            }
+        }
+
+        /// <summary>
+        ///     Whether there's a midi song being played or not.
+        /// </summary>
+        [ViewVariables]
+        public bool IsMidiOpen => _renderer.Status == MidiRendererStatus.File;
+
+        /// <summary>
+        ///     Whether the midi renderer is listening for midi input or not.
+        /// </summary>
+        [ViewVariables]
+        public bool IsInputOpen => _renderer.Status == MidiRendererStatus.Input;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            IoCManager.InjectDependencies(this);
+            _renderer = _midiManager.GetNewRenderer();
+            _renderer.MidiProgram = _instrumentProgram;
+            _renderer.TrackingEntity = Owner;
+            _renderer.OnMidiPlayerFinished += () => { OnMidiPlaybackEnded?.Invoke(); };
+        }
+
+        protected override void Shutdown()
+        {
+            base.Shutdown();
+            _renderer?.Dispose();
+        }
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _instrumentProgram, "program", 1);
+        }
+
+        public override void HandleMessage(ComponentMessage message, INetChannel netChannel = null, IComponent component = null)
+        {
+            base.HandleMessage(message, netChannel, component);
+            switch (message)
+            {
+                case InstrumentMidiEventMessage midiEventMessage:
+                    // If we're the ones sending the MidiEvents, we ignore this message.
+                    if (IsInputOpen || IsMidiOpen) break;
+                    _renderer.SendMidiEvent(midiEventMessage.MidiEvent);
```

Comment:
I don't see anything ensuring correct timing on the MIDI playback here, meaning this will be *incredibly* prone to notes getting mistimed, especially over poorer network connections.

Reply:
Hmm, right... Do you have any ideas to fix this?

Reply:
You're gonna want to send some sort of timing information with each note, so the client can play them exactly when they should. You'll also want to delay all note playback by some value like 200ms so is some packet gets dropped it can be re-sent in time.

Reply:
This *should* be fine now.